File: samples/bot-meeting-ai-insights/python/dialogs/main_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core import MessageFactory
from botbuilder.dialogs import (
    WaterfallDialog,
    WaterfallStepContext,
    DialogTurnResult,
    PromptOptions,
)
from botbuilder.dialogs.prompts import OAuthPrompt, OAuthPromptSettings, ConfirmPrompt

from dialogs import LogoutDialog
import logging
import aiohttp
import os
import json
from urllib.parse import quote

# Set the logging level to INFO
logging.basicConfig(level=logging.INFO)

class MainDialog(LogoutDialog):
    def __init__(self, connection_name: str):
        # Initializes the MainDialog with OAuthPrompt and WaterfallDialog.
        super(MainDialog, self).__init__(MainDialog.__name__, connection_name)

        self.add_dialog(
            OAuthPrompt(
                OAuthPrompt.__name__,
                OAuthPromptSettings(
                    connection_name=connection_name,
                    text="Please Sign In",
                    title="Sign In",
                    timeout=300000
                )
            )
        )

        self.add_dialog(ConfirmPrompt(ConfirmPrompt.__name__))

        self.add_dialog(
            WaterfallDialog(
                "WFDialog",
                [
                    self.prompt_step,
                    self.login_step,
                ],
            )
        )

        self.initial_dialog_id = "WFDialog"

    # Prompts the user to sign in using the OAuthPrompt.
    async def prompt_step(self, step_context: WaterfallStepContext) -> DialogTurnResult:
        logging.info("Reached OAuthPrompt step — prompting for login.")
        return await step_context.begin_dialog(OAuthPrompt.__name__)

    # Handles the login result and retrieves AI insights.
    async def login_step(self, step_context: WaterfallStepContext) -> DialogTurnResult:
        token_response = step_context.result
        if token_response and token_response.token:
            sso_token = token_response.token
            
            if sso_token:
                user_id = os.environ.get('UserId')
                meeting_url = os.environ.get('MeetingJoinUrl')
                
                online_meeting_id = await self.get_meeting_id(sso_token, meeting_url)
                ai_insight_id = await self.get_ai_insight_id(sso_token, user_id, online_meeting_id)
                
                if ai_insight_id:
                    ai_insight = await self.get_ai_insight_details(sso_token, user_id, online_meeting_id, ai_insight_id)
                    if ai_insight:
                        formatted_message = ''
                        if isinstance(ai_insight, list):
                            formatted_message = '\n'.join([
                                f"## {insight.get('title', '')}\n{insight.get('text', '')}\n"
                                for insight in ai_insight
                            ])
                        
                        await step_context.context.send_activity(
                            formatted_message or 'No insights found in the expected format.'
                        )
                    else:
                        await step_context.context.send_activity('Failed to retrieve AI Insight details.')
                else:
                    await step_context.context.send_activity('Failed to retrieve AI Insight ID.')
            else:
                await step_context.context.send_activity('Failed to retrieve access token for downstream API.')
            
            return await step_context.end_dialog()
        
        await step_context.context.send_activity('Login was not successful, please try again.')
        return await step_context.end_dialog()

    async def get_meeting_id(self, access_token: str, meeting_url: str) -> str:
        """
        Calls the Graph API to get the meeting ID from the meeting URL.
        
        Args:
            access_token: The access token for the Graph API.
            meeting_url: The meeting join URL.
            
        Returns:
            The online meeting ID or None if failed.
        """
        try:
            url = f"https://graph.microsoft.com/beta/me/onlineMeetings?$filter=JoinWebUrl eq '{meeting_url}'"
            
            headers = {
                'Authorization': f'Bearer {access_token}'
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(url, headers=headers) as response:
                    if response.status == 200:
                        data = await response.json()
                        if data.get('value') and len(data['value']) > 0:
                            return data['value'][0]['id']
                    else:
                        logging.error(f"Error retrieving meeting ID: {response.status}")
                        
        except Exception as error:
            logging.error(f'Error retrieving meeting ID: {error}')
            
        return None

    async def get_ai_insight_id(self, access_token: str, user_id: str, online_meeting_id: str) -> str:
        """
        Calls the Graph API to get the AI Insight ID.
        
        Args:
            access_token: The access token for the Graph API.
            user_id: The user ID.
            online_meeting_id: The online meeting ID.
            
        Returns:
            The AI Insight ID or None if failed.
        """
        try:
            url = f"https://graph.microsoft.com/beta/copilot/users/{user_id}/onlineMeetings/{online_meeting_id}/aiInsights"
            
            headers = {
                'Authorization': f'Bearer {access_token}'
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(url, headers=headers) as response:
                    if response.status == 200:
                        data = await response.json()
                        if data.get('value') and len(data['value']) > 0:
                            return data['value'][0]['id']
                    else:
                        logging.error(f"Error retrieving AI Insight ID: {response.status}")
                        
        except Exception as error:
            logging.error(f'Error retrieving AI Insight ID: {error}')
            
        return None

    async def get_ai_insight_details(self, access_token: str, user_id: str, online_meeting_id: str, ai_insight_id: str):
        """
        Calls the Graph API to get the AI Insight details.
        
        Args:
            access_token: The access token for the Graph API.
            user_id: The user ID.
            online_meeting_id: The online meeting ID.
            ai_insight_id: The AI Insight ID.
            
        Returns:
            The meeting notes from AI insights or None if failed.
        """
        try:
            url = f"https://graph.microsoft.com/beta/copilot/users/{user_id}/onlineMeetings/{online_meeting_id}/aiInsights/{ai_insight_id}"
            
            headers = {
                'Authorization': f'Bearer {access_token}'
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(url, headers=headers) as response:
                    if response.status == 200:
                        data = await response.json()
                        return data.get('meetingNotes')
                    else:
                        logging.error(f"Error retrieving AI Insight details: {response.status}")
                        
        except Exception as error:
            logging.error(f'Error retrieving AI Insight details: {error}')
            
        return None

File: samples/TeamsSDK/Archived/bot-meeting-ai-insights/nodejs/dialogs/mainDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ConfirmPrompt, DialogSet, DialogTurnStatus, OAuthPrompt, WaterfallDialog } = require('botbuilder-dialogs');
const { LogoutDialog } = require('./logoutDialog');
const axios = require('axios');

const CONFIRM_PROMPT = 'ConfirmPrompt';
const MAIN_DIALOG = 'MainDialog';
const MAIN_WATERFALL_DIALOG = 'MainWaterfallDialog';
const OAUTH_PROMPT = 'OAuthPrompt';

/**
 * MainDialog class that extends LogoutDialog to handle the main dialog flow.
 */
class MainDialog extends LogoutDialog {
    /**
     * Creates an instance of MainDialog.
     * @param {string} id - The dialog ID.
     * @param {string} connectionName - The connection name for the OAuth provider.
     */
    constructor() {
        super(MAIN_DIALOG, process.env.connectionName);

        this.addDialog(new OAuthPrompt(OAUTH_PROMPT, {
            connectionName: process.env.connectionName,
            text: 'Please Sign In',
            title: 'Sign In',
            timeout: 300000,
        }));
        this.addDialog(new ConfirmPrompt(CONFIRM_PROMPT));
        this.addDialog(new WaterfallDialog(MAIN_WATERFALL_DIALOG, [
            this.promptStep.bind(this),
            this.loginStep.bind(this)
        ]));

        this.initialDialogId = MAIN_WATERFALL_DIALOG;
    }

    /**
     * The run method handles the incoming activity (in the form of a DialogContext) and passes it through the dialog system.
     * If no dialog is active, it will start the default dialog.
     * @param {TurnContext} context - The context object for the turn.
     * @param {StatePropertyAccessor} accessor - The state property accessor for the dialog state.
     */
    async run(context, accessor) {
        const dialogSet = new DialogSet(accessor);
        dialogSet.add(this);

        const dialogContext = await dialogSet.createContext(context);
        const results = await dialogContext.continueDialog();
        if (results.status === DialogTurnStatus.empty) {
            await dialogContext.beginDialog(this.id);
        }
    }

    /**
     * Prompts the user to sign in.
     * @param {WaterfallStepContext} stepContext - The waterfall step context.
     */
    async promptStep(stepContext) {
        return await stepContext.beginDialog(OAUTH_PROMPT);
    }

    /**
     * Handles the login step and retrieves AI insights.
     * @param {WaterfallStepContext} stepContext - The waterfall step context.
     */
    async loginStep(stepContext) {
        const tokenResponse = stepContext.result;
        if (tokenResponse) {
            const ssoToken = tokenResponse.token;

            if (ssoToken) {
                const userId = process.env.userId;
                const meetingUrl = process.env.meetingJoinUrl;
                const onlineMeetingId = await this.getMeetingId(ssoToken, meetingUrl);
                const aiInsightId = await this.getAiInsightId(ssoToken, userId, onlineMeetingId);

                if (aiInsightId) {
                    const aiInsight = await this.getAiInsightDetails(ssoToken, userId, onlineMeetingId, aiInsightId);                    if (aiInsight) {
                        let formattedMessage = '';
                        if (Array.isArray(aiInsight)) {
                            formattedMessage = aiInsight.map(insight => {
                                return `## ${insight.title}\n${insight.text}\n`;
                            }).join('\n');
                        }
                        await stepContext.context.sendActivity(formattedMessage || 'No insights found in the expected format.');
                    } else {
                        await stepContext.context.sendActivity('Failed to retrieve AI Insight details.');
                    }
                } else {
                    await stepContext.context.sendActivity('Failed to retrieve AI Insight ID.');
                }
            } else {
                await stepContext.context.sendActivity('Failed to retrieve access token for downstream API.');
            }
            return await stepContext.endDialog();
        }
        await stepContext.context.sendActivity('Login was not successful, please try again.');
        return await stepContext.endDialog();
    }

    async getMeetingId(accessToken, meetingUrl) {
        try {
            const url = `https://graph.microsoft.com/v1.0/me/onlineMeetings?$filter=JoinWebUrl eq '${meetingUrl}'`;

            const response = await axios.get(url, {
                headers: {
                    Authorization: `Bearer ${accessToken}`
                }
            });
            return response.data.value[0].id;
        } catch (error) {
            console.error('Error retrieving AI Insight ID:', error);
            return null;
        }
    }

    /**
     * Calls the Graph API to get the AI Insight ID.
     * @param {string} accessToken - The access token for the Graph API.
     * @param {string} userId - The user ID.
     * @param {string} onlineMeetingId - The online meeting ID.
     * @returns {Promise<string>} - The AI Insight ID.
     */
    async getAiInsightId(accessToken, userId, onlineMeetingId) {
        try {
            const url = `https://graph.microsoft.com/beta/copilot/users/${userId}/onlineMeetings/${onlineMeetingId}/aiInsights`;

            const response = await axios.get(url, {
                headers: {
                    Authorization: `Bearer ${accessToken}`
                }
            });
            return response.data.value[0].id;
        } catch (error) {
            console.error('Error retrieving AI Insight ID:', error);
            return null;
        }
    }

    /**
     * Calls the Graph API to get the AI Insight details.
     * @param {string} accessToken - The access token for the Graph API.
     * @param {string} userId - The user ID.
     * @param {string} onlineMeetingId - The online meeting ID.
     * @param {string} aiInsightId - The AI Insight ID.
     */
    async getAiInsightDetails(accessToken, userId, onlineMeetingId, aiInsightId) {
        try {
            const url = `https://graph.microsoft.com/beta/copilot/users/${userId}/onlineMeetings/${onlineMeetingId}/aiInsights/${aiInsightId}`;

            const response = await axios.get(url, {
                headers: {
                    Authorization: `Bearer ${accessToken}`
                }
            });

            return response.data.meetingNotes;
        } catch (error) {
            console.error('Error retrieving AI Insight details:', error);
            return null;
        }
    }
}

module.exports.MainDialog = MainDialog;